core/src/main/java/com/github/mizool/core/concurrent/SynchronizerApi.java

```diff
@@ -20,16 +21,37 @@ interface SleepRunGet extends RunGet
         /**
          * Adds sleeping to the action chain. <br>
          * <br>
-         * When performing this action, the chain will sleep until another chain wakes it up, then call the given
-         * supplier. If the supplier returns {@code false}, this action chain resumes sleeping. Otherwise, the chain
-         * continues by performing the main action.
+         * When performing this action, the chain will ensure the given condition is met before performing the main
+         * action. If the supplier returns {@code false}, this action chain sleeps until another chain wakes it up.
+         * Then, the supplier is called again and if it still returns {@code false}, the chain resumes sleeping.
+         * Otherwise, the chain continues by performing the main action.
          *
          * @param state the supplier that returns {@code true} if the action chain should continue, {@code false}
          * otherwise.
          *
          * @throws NullPointerException if {@code state} is null
          */
-        RunGet sleepUntil(BooleanSupplier state);
+        default RunGet sleepUntil(BooleanSupplier state)
+        {
+            return sleepUntil(state, null);
+        }
+
+        /**
+         * Adds sleeping to the action chain. <br>
+         * <br>
+         * When performing this action, the chain will ensure the given condition is met before performing the main
+         * action. If the supplier returns {@code false}, this action chain sleeps until either another chain wakes it
+         * up or the timeout has passed. Then, the supplier is called again and if it still returns {@code false}, the
+         * chain resumes sleeping. Otherwise, the chain continues by performing the main action.
+         *
+         * @param state the supplier that returns {@code true} if the action chain should continue, {@code false}
+         * otherwise.
+         * @param timeout the maximum amount of time to sleep before re-checking the condition, or {@code null} to
+         * sleep indefinitely.
+         *
+         * @throws NullPointerException if {@code state} is null
+         */
+        RunGet sleepUntil(BooleanSupplier state, Duration timeout);
```

Comment:
Would it make sense to exchange the boolean supplier for a function that receives the information about whether the timeout was reached or another thread was waking it up? Something like `Function<WakeReason, Boolean>` perhaps?

Since the second parameter is called `timeout`, I would imagine that implementations of `state` might want to act differently, depending on whether the process was resumed "normally" or went into a timeout.

Reply:
@bannmann I've approved and the pull request as I don't see this as something we _have_ to do but more as a conversation starter.

If you want' we can talk about this. But "yagni" is also fine. ;)

Reply:
For the record: we discussed this in person and the tl;dr of our conclusion is YAGNI.